### ISSUE TITLE
Use PROJ for CRS transformations instead of Geotools

### ISSUE THREAD

**Issue by cportele** [interactive-instruments/ldproxy#439]
An initial implementation that mostly works was done in #512. But intensive tests showed that PROJ-JNI is not as production-ready as we hoped:

- [x] shutdown due to out of memory exception in PROJ_JNI for certain CRSs, with an underlying file not found error in PROJ, most likely due to a missing grid file (also see https://github.com/interactive-instruments/ldproxy-project-management/issues/129#issuecomment-988712640 ff.)
- [x] memory leak when transformer/operation caching is disabled (see https://github.com/interactive-instruments/ldproxy-project-management/issues/129#issuecomment-988733272). 
- [x] `point outside of projection domain` errors with transformations results of `infinity`, where GeoTools returned at least some however nonsensical values, has to be handled in higher layers like tile generation
  - I think we can leave this like it is, it should not occur anymore when #594 is implemented. The error only occurs for really abnormal values. In the case of lika the affected coordinate is near Sumatra, and a point slightly northwards of that works as expected.
- [x] add configuration option for PROJ configuration & data directory, or at least allow to use the default behavior how PROJ locates the resource directories (https://proj.org/resource_files.html). Also make sure that ldproxy will not overwrite PROJ 
  - I added an option to `cfg.yml`:
    ```yml
    proj:
      location: /home/zahnen/proj
    ```
- [x] GeoTIFF support
- [ ] support RHEL7

Furthermore we identified some follow-up issues regarding the `xtraplatform-crs` api:

- [x] 3D transformers cannot handle 2D coordinates
- [x] cleanup api + implementation, e.g. `CoordinateTupleWithPrecision`, `CrsTransformerProj.isSourceMetric`, `CrsTransformerProj.isTargetMetric`
  - moved to #597

**Comment by azahnen** [comment]
@bekakh 
The first task would be to compile PROJ for Linux. So you would have to find out how that works in your WSL environment. The instructions can be found here: https://kortforsyningen.github.io/PROJ-JNI/install.html

If you are able to compile PROJ and the test from step 4 of the instructions succeeds, you could then go on and implement PROJ variants of `GeoToolsCrsTransformerFactory` and `GeoToolsCrsTransformer` in `xtraplatform-crs`. You would have to use your local PROJ-JNI jar as a file dependency in `build.gradle`, see https://docs.gradle.org/current/userguide/declaring_dependencies.html#sub:file_dependencies.

I would then automate the build of PROJ and PROJ-JNI later, so that it can be used as a normal dependency.

**Comment by azahnen** [comment]
@cportele @rkoblenzer I updated the docker image. I checked/commented the resolved issues above.

**Comment by cportele** [comment]
I did some tests with the features and tiles resources of the demo and lika configurations. Here are things that did not work as expected so far:

* I had to disable the zoomstack API, because the OSGB CRS requires a gridshift file. While it accepted a GSB file for Beta2007 instead of a GeoTIFF file, this did not work for the OSGB CRS.
* Tile seeding ran for a few hours in the lika API, but then suddenly stopped, without any hint about this in the log (level: DEBUG). I will check, if the behaviour can be reproduced tonight.

**Comment by cportele** [comment]
I ran some additional tests with the seeding. There seems to be an issue, but my guess is that it is not related to the CRS part and I should transfer this to its own issue. Seeding of lika with 8 threads. The seeding runs for a while and suddenly only two of the 8 threads are active, there are no tiles reported for any of the other 6 threads, but also no information about any error or termination of a thread. There was an SQL I/O error reported for one tile, so this may be related.

```
... all threads 1 to 8
DEBUG [2022-02-04 16:25:53,407]                     lika - Tile cache seeding: currently processing -> multi-layer, WebMercatorQuad/14/5455/8585, pbf [16%] [2/8]  
DEBUG [2022-02-04 16:25:54,366]                     lika - Tile cache seeding: processing failed -> multi-layer, WebMercatorQuad/14/5456/8544, pbf | Unexpected SQL query error: An I/O error occurred while sending to the backend.  
DEBUG [2022-02-04 16:25:54,366]                     lika - Tile cache seeding: currently processing -> multi-layer, WebMercatorQuad/14/5456/8544, pbf [16%] [1/8]  
DEBUG [2022-02-04 16:25:55,375]                     lika - Tile cache seeding: currently processing -> multi-layer, WebMercatorQuad/13/2740/4277, pbf [5%] [6/8]  
DEBUG [2022-02-04 16:25:58,291]                     lika - Tile cache seeding: currently processing -> multi-layer, WebMercatorQuad/14/5456/8552, pbf [16%] [1/8]  
DEBUG [2022-02-04 16:26:03,674]                     lika - Tile cache seeding: currently processing -> multi-layer, WebMercatorQuad/13/2740/4285, pbf [5%] [6/8]  
... only threads 1 and 6
```

**Comment by azahnen** [comment]
@cportele @rkoblenzer I updated the image, it should now support GeoTIFF grid files.

**Comment by rkoblenzer** [comment]
@azahnen , thanks. I get a stracktrace with the updated image (it did not occur yesterday): 

```
INFO  [2022-02-06 16:19:42,855]                          - Starting ldproxy 3.1.0-SNAPSHOT
ERROR [2022-02-06 16:19:45,498]                          - PROJ: Could not load library: /ldproxy/data/tmp/proj-8.2.0-1/linux-x86_64/libproj.so
DEBUG [2022-02-06 16:19:45,500]                          - Stacktrace:
! java.lang.IllegalStateException: Could not load library: /ldproxy/data/tmp/proj-8.2.0-1/linux-x86_64/libproj.so
! at de.ii.xtraplatform.nativ.XtraplatformNative.loadLib(XtraplatformNative.java:86)
...
```
My guess: ldd is reporting (within the docker container) missing libs:

```
root@76fb50f0e3e9:/ldproxy# LD_LIBRARY_PATH=/ldproxy/data/tmp/proj-8.2.0-1/linux-x86_64 ldd /ldproxy/data/tmp/proj-8.2.0-1/linux-x86_64/libproj.so
 ...
libsqlite3.so.0 => not found
libtiff.so.5 => not found
```
It seems that the symlinks are missing:

```
root@76fb50f0e3e9:/ldproxy# ls /ldproxy/data/tmp/proj-8.2.0-1/linux-x86_64
libproj.so  libsqlite3.so  libtiff.so  libz.so
``` 

Note: I removed already the temporary directories (felix and proj-8.2..). 

BTW, is the new parameter `proj` in `cfg.yml` a top-level parameter?

**Comment by azahnen** [comment]
@rkoblenzer Sorry, the libs were loaded in the wrong order, please try again. Yes, `proj` is a top-level property.

**Comment by rkoblenzer** [comment]
@azahnen , no problem, it works now - thanks.

**Comment by cportele** [comment]
@azahnen - thanks, the error message about the missing gridshift file is now gone in the zoomstack API.

However, I am now getting an error during hydration when a computed bounding box in the native OSGB CRS is converted to CRS84 (all geometries are in GB). I get:

> Error transforming the bounding box with CRS 'http://www.opengis.net/def/crs/EPSG/0/27700'

and 

> org.opengis.referencing.operation.TransformException: Coordinate to transform falls outside grid

**Comment by cportele** [comment]
I have tested with 3D data and it works as expected.

**Comment by azahnen** [comment]
@cportele I found out that which operation is chosen by PROJ can be affected via parameters, and that with the default values it always chooses the grid file based operation for EPSG:27700, even if the grid file is missing and reasonable alternative operations would be available. But by adjusting the parameters I can bring it to use the grid file based operation when the file is available and the best alternative otherwise.
I adjusted the parameters and updated the image, so when you remove the grid file, zoomstack should work without errors.
To implement a workaround for the error that occurs with grid file would require bigger changes to the C++ layer in PROJ-JNI, I created a follow-up task for that: #596

@rkoblenzer The above also applies to lika and EPSG:31466. That means instead of throwing an error when the beta grid file is missing, it should now use a slightly less accurate operation.
